keps/sig-apps/2185-random-pod-select-on-replicaset-downscale/README.md

```diff
@@ -0,0 +1,437 @@
+   Pods.
+6. ReplicaSet is downscaled to 2N again. Due to the downscaling preference, all
+   the Pods from one domain are removed, leading to imbalance. The situation
+   doesn't improve with repeated upscale and downscale steps. Instead, a
+   randomized approach leaves about 2/3*N nodes in each
+   failure domain.
+
+### Notes/Constraints/Caveats
+
+The original heuristic, dowscaling the youngest Pods first, has its benefits.
+Newer Pods might not have finished starting up (or warming up) and are likely
+to have less active connections than older Pods. However, this distinction
+doesn't generally apply once Pods have been running steadily for some time.
+
+A purely randomized approach would break those assumptions, potentially
+leading to services disruption. Choosing a heuristic could be left to the user.
+On the other hand, certain workloads take a long time to warm up and, at the
+same time, require high availability.
+
+### Risks and Mitigations
+
+Certain users might be relaying in the existing downscaling heuristic. However,
+there are a number of reasons why we don't need to preserve such behavior as is:
+
+- The behavior is not documented in the API or user docs, only in code.
+- The heuristic is applied last, after other criteria have been applied. In
+  particular, the heuristic doesn't apply when multiple Pods are in the same
+  Node or if a particular Pod has had several container restarts. While users
+  can enforce one Pod per Node with certain scheduling features, there is no
+  workaround for other criteria.
+- We are not proposing to entirely remove the heuristic, just make it more
+  laxed.
+- We are introducing a [related feature](git.k8s.io/enhancements/keps/sig-apps/1828-delete-priority-annotations)
+  that provides higher guarantees for downscaling order that users can migrate
+  to.
+  
+  <!--
+  If not merged yet, the KEP can be found in https://github.com/kubernetes/enhancements/pull/1828
+  -->
+
+## Design Details
+
+We propose a randomized approach to the algorithm for Pod victim selection
+during ReplicaSet downscale:
+
+1. Do a random shuffle of ReplicaSet Pods.
+2. Obtain wall time, and add it to [`ActivePodsWithRanks`](https://github.com/kubernetes/kubernetes/blob/dc39ab2417bfddcec37be4011131c59921fdbe98/pkg/controller/controller_utils.go#L815)
+2. Call sorting algorithm with a modified time comparison for start and
+   creation timestamp.
+
+Instead of directly comparing timestamps, the algorithm compares the elapsed
+times since the timestamp until the current time but in a logarithmic scale,
+floor rounded. This has the effect of treating elapsed times as equals when they
+have the same scale. That is, Pods that have been running for a few nanoseconds
+are equal, but they are different from pods that have been running for a few
+seconds or a few days.
+
+For example, let's assume the base 10 is used, then we have the following
+mapping for different durations:
+
+| Duration | Scale |
```

Comment:
perhaps clarify that scale translates to the pod rank.

Reply:
rank as defined in the code? It doesn't

Reply:
Not literally as defined in the code, but that it translates to the order.

Reply:
It does not. It just changes how timestamps are compared, which is the last criteria for Pod comparisons.

Reply:
ok, may be I am being too brief since I am commenting from the phone (you also need to give me the benefit of the doubt :)), what I meant is that this is the new sorting key.

Reply:
Hopefully this gets explained with the other comment.
It can be seen as a sorting key, but saying that would create confusion, as it is not the only criteria.

Reply:
there are multiple sorting keys, not just one.